Parse doc string from sourcecode

t-wissmann opened the pull request:
This contributes the following:

  - A new C++ base class HasDocumentation is added. It has a member
    function setDoc() which allows to set user documentation within the
    hlwm source code. So far, there is no functionality to pass the doc
    string further to the user
  - doc/gendoc.py is extended such that it reads these invocation of
    setDoc() and puts this documentation text into the hlwm-doc.json

blarz commented on doc/gendoc.py:
```diff
@@ -656,6 +672,20 @@ def stream_consume_member_initializers(self, classname, stream):
+        # we're now on a codeblock
+        self.stream_consume_setDoc(classname, TokenStream(codeblock.value.enclosed_tokens))
+
+    def stream_consume_setDoc(self, classname, stream):
```
Should test_gendoc.py be extended for the whole doc parsing thing?

t-wissmann replied:
There is no way to obtain the doc string from a running hlwm instance, so we can't test it. Once there is a `help` command, I planned to test it.

blarz replied:
I see, thanks!

blarz commented on src/tag.cpp:
```diff
@@ -57,6 +57,15 @@ HSTag::HSTag(string name_, TagManager* tags, Settings* settings)
         return this->floatingLayerCanBeFocused(v);
     });
+
+    floating.setDoc("if the entire tag is set to floating mode");
```
For me it feels weird to extend the code by a new class in order to extend documentation/generated JSON.

t-wissmann replied:
In the end, I'd like to have a command `help` with the following behaviour:
```
$ hc help tags.0.floating
type: boolean setting
current value: false
meaning: if the entire tag is set to floating mode
```
https://github.com/herbstluftwm/herbstluftwm/pull/998

blarz replied:
My initial dislike was directed to the latter, but seeing the example of the help command, it makes much more sense to me now. I didn't realize the potential of that plan in the commit message of other PR.

I was just confused because normally "documentation" of this sort was kind of generated by doxygen or doxygen comments, but I know that this is not something we can use here.

So I'm gonna approve this (after a few minor typos are fixed) and see where this goes :)

blarz reviewed (changes requested):
Two minor things. Also the commented out setDoc call in frametree.cpp can be enabled, right?

t-wissmann commented:
I've removed the comment, added the setDoc and fixed the typo.

Unfortunatelly, I didn't manage to accomplish this documentation generation with doxygen. Doxygen seems to work well for api documentation but less well for user documentation.

blarz commented:
> I've removed the comment, added the setDoc and fixed the typo.

Thanks!

> Unfortunatelly, I didn't manage to accomplish this documentation generation with doxygen. Doxygen seems to work well for api documentation but less well for user documentation.

I know, it's just my goto source documentation software I guess.
But your approach is interesting, let's see how it works out :)
